### DISCUSSION TITLE
Status Update - 1 Billion Keys!

### DISCUSSION THREAD

**Discussion by koculu** [koculu/ZoneTree#9] [Announcements]
Starting with [v1.4.4](https://github.com/koculu/ZoneTree/releases/tag/release-v1.4.4),
ZoneTree supports "**Vertical Disk Segment Expansion**" in addition to the "**Horizontal Disk Segment Expansion**".
That enables the insertion of billions or trillions of data with stable performance.

Please see the 1 billion integer key-value insert benchmark below.
241 seconds to insert,
150 seconds to iterate
🥇 1 Billion keys. 🥇 

![image](https://user-images.githubusercontent.com/7453467/187039553-e91d5dff-0a5a-477d-ad60-e47c9c5e2032.png)

### Future note: 
The ideal bottom segment layer is a pyramid-like shape.
I work on a new background operation feature to optimize the bottom segment layer's shape.

Keep in touch...

**Comment by jihadkhawaja** [comment]
Impressive 👏🏻

**Comment by koculu** [comment]
The version [1.4.5](https://github.com/koculu/ZoneTree/releases/tag/release-v1.4.5)
introduces the Bottom Segment Merger. 
[API Docs: https://tenray.io/docs/ZoneTree/api/Tenray.ZoneTree.IZoneTreeMaintenance-2.html](https://tenray.io/docs/ZoneTree/api/Tenray.ZoneTree.IZoneTreeMaintenance-2.html#Tenray_ZoneTree_IZoneTreeMaintenance_2_StartBottomSegmentsMergeOperation_System_Int32_System_Int32_) 

Users now can control the bottom segments layout by merging with new API and events.
https://github.com/koculu/ZoneTree/releases/tag/release-v1.4.5

The following picture shows a billion length database layout after manual bottom segment merges applied.
![image](https://user-images.githubusercontent.com/7453467/187099650-ff87f6cd-9af1-4186-bfb9-f38b413a4341.png)
